Support running with semi-sync enforced.

enisoc opened the pull request:
That is, with async fallback not allowed under any circumstance for
master writes. This is accomplished, for example, by setting:

```
rpl_semi_sync_master_wait_no_slave = 1
rpl_semi_sync_master_timeout = 1000000000000000000 # "infinite"
```

When Orchestrator works with instances that use these settings, it's
important to toggle `rpl_semi_sync_master_enabled` and
`rpl_semi_sync_slave_enabled` at the right times to ensure that:

* A master must not be not allowed to accept writes until semi-sync is
  turned on.
* Slaves don't get stuck waiting for ACKs when they have no slaves of
  their own.
* Slaves with `PromotionRule: must_not` must never ACK, because if they
  win the ACK race when the master goes down, we're stuck.

Orchestrator will use the `DetectSemiSyncEnforcedQuery` config entry
(if specified) to determine if a given instance is configured in this
mode. As an example, you can set `DetectSemiSyncEnforcedQuery` to
something like this:

```
SELECT @@global.rpl_semi_sync_master_wait_no_slave AND @@global.rpl_semi_sync_master_timeout > 1000000
```

See #156 for more information.

shlomi-noach commented on go/inst/instance_topology_dao.go:
```diff
@@ -349,6 +350,20 @@ func StartSlave(instanceKey *InstanceKey) (*Instance, error) {
+			return instance, log.Errore(err)
+		}
+	}
+
```
Perhaps I'm not familiar enough with semi-sync; but why is it necessary to `set global rpl_semi_sync_master_enabled = ?, global rpl_semi_sync_slave_enabled = ?` upon each and every `START SLAVE`?

enisoc replied:
We set both `master_enabled` and `slave_enabled` to the proper values explicitly, so they're guaranteed to be right regardless of any prior history of that node. Enforcing this invariant makes it easy to reason about the risks of alternate futures or stuck slaves.

The reason we do it before `START SLAVE` in particular is because changes to the `slave_enabled` variable only take effect at the time of `START SLAVE`. If changed at any other time, the only way to ensure it takes effect would be to `STOP SLAVE; START SLAVE`.

shlomi-noach replied:
:+1: I've actually hit the exact same use case in another project.

enisoc commented:
I've moved schema changes to the end of the list as mentioned in #194. I've also removed the WIP label as I consider it ready whenever you agree, although I will likely have to fix merge conflicts in `db.go` and `instance_dao.go` in whichever PR goes second.
